src/plugins/compressors/bg_plugin.cc

```diff
@@ -0,0 +1,128 @@
+      set(options, "bg:errorControlMode", errorControlMode_libpressio);
+      set(options, "bg:nsd", nsd_libpressio);
+      set(options, "bg:dsd", dsd_libpressio);
+      return options;
+    }
+
+    struct pressio_options get_configuration_impl() const override {
+      struct pressio_options options;
+      options.set("pressio:thread_safe", static_cast<int>(pressio_thread_safety_multiple));
+      return options;
+    }
+
+    int set_options_impl(struct pressio_options const& options) override {
+      get(options, "bg:bgMode", &bgMode_libpressio);
+      get(options, "bg:errorControlMode", &errorControlMode_libpressio);
+      get(options, "bg:nsd", &nsd_libpressio);
+      get(options, "bg:dsd", &dsd_libpressio);
+      return 0;
+    }
+
+    int compress_impl(const pressio_data *input, struct pressio_data* output) override {
+      int type = libpressio_type_to_bg_type(pressio_data_dtype(input));
+      if(type == -1) {
+         return set_error(2, "Invalid data type");
+      }
+
+      size_t nbEle = static_cast<size_t>(pressio_data_num_elements(input));
+      unsigned long outSize;
+      void* data = pressio_data_ptr(input, nullptr);
+      unsigned char* compressed_data;
+
+
+      compressed_data = BG_compress_args(type, data, &outSize, bgMode_libpressio, errorControlMode_libpressio, nsd_libpressio, dsd_libpressio, nbEle);
+
+      //that means the compressor is complaining about the parameter
+      if(compressed_data == NULL)
+      {
+        return set_error(2, "Error when bg is compressing the data");
+      }
+
+      *output = pressio_data::move(pressio_byte_dtype, compressed_data, 1, &outSize, pressio_data_libc_free_fn, nullptr);
+      return 0;
+    }
+
+    int decompress_impl(const pressio_data *input, struct pressio_data* output) override {
+      int type = libpressio_type_to_bg_type(pressio_data_dtype(output));
+      if(type == -1) {
+         set_error(2, "Invalid data type");
+      }
+      unsigned char* bytes = (unsigned char*)pressio_data_ptr(input, nullptr);
+      size_t nbEle = static_cast<size_t>(pressio_data_num_elements(output));
+      size_t byteLength = pressio_data_get_bytes(input);
+
+      void* decompressed_data = BG_decompress(type, bytes, byteLength, nbEle);
+      *output = pressio_data::move(pressio_data_dtype(output), decompressed_data, 1, &nbEle, pressio_data_libc_free_fn, nullptr);
+      return 0;
+      }
+
+
+    
+    int major_version() const override {
+      return 0;
+    }
+    int minor_version() const override {
+      return 0;
+    }
+    int patch_version() const override {
+      return 0;
+    }
+    int revision_version () const { 
+      return 1;
+    }
+
+    const char* version() const override {
+      return "0.0.1"; 
+    }
+
+
+    const char* prefix() const override {
+      return "bg";
+    }
+
+    std::shared_ptr<libpressio_compressor_plugin> clone() override {
+      return compressor_plugins().build("bg");
```

Comment:
This doesn't follow the correct semantics.  clone() should also copy the configuration parameters passed by the user to earlier calls to set_options.  An easy way to do this is to call

```cpp
return compat::make_unique<bit_grooming_plugin>(*this);
```

which invokes the copy constructor